templates/master/00-master/on-prem/files/keepalived-script-both.yaml

```diff
@@ -3,4 +3,4 @@ path: "/etc/kubernetes/static-pod-resources/keepalived/scripts/chk_ocp_script_bo
 contents:
   inline: |
     #!/bin/bash
-    /usr/bin/curl -o /dev/null -kLfs https://localhost:{{`{{ .LBConfig.LbPort }}`}}/readyz && [ -e /var/run/keepalived/iptables-rule-exists ] || /usr/bin/curl -kLfs https://localhost:{{`{{ .LBConfig.ApiPort }}`}}/readyz
+    /usr/bin/curl -o /dev/null -kLfs https://localhost:9444/haproxy_ready && [ -e /var/run/keepalived/iptables-rule-exists ] || /usr/bin/curl -kLfs https://localhost:{{`{{ .LBConfig.ApiPort }}`}}/readyz
```

Comment:
The concern I have here is that it doesn't actually verify that the api is reachable through haproxy. Just because haproxy is running that doesn't mean everything is working correctly. For example, I'm not sure that on startup we require the haproxy backend list to be populated, so it's possible it could be running with nothing behind it, and if it takes the vip in that state then we'll never recover.

TLDR: the fact that this hits an api endpoint was deliberate.

That said, I'm not immediately coming up with a good solution to the scenario you describe. Maybe some sort of backoff so that if the VIP moves a lot in a short period of time we get less aggressive with our health checks? Or just if we detect instability in the api? That might be easier to implement since we could keep track of how many api requests are failing in the monitor, and if we see more than X% failures we increase the fall value in the health check so the vip is less likely to move due to flakes.

Reply:
Thanks for the feedback. I'm gonna have to study exactly how the pivot from bootstrap works to see if we can end up in the deadlock situation you're describing. Stay tuned.

Reply:
So, I had a look at the bootstrap sequence and I don't think it's going to be a problem.
The API VIP remains on the bootstrap node until the bootstrap node is torn down, since https://github.com/openshift/machine-config-operator/commit/1537a2234580fea0d4e17bbb5b2e998c09cfca4f.
At that point, we have:
- masters have all booted and started their static pods
- the etcd cluster formed with master nodes is healthy
- the bootstrap control plane is up
- cluster-etcd-operator is healthy

For both haproxy and keepalived unicast, we get the list of backends from baremetal-runtimecfg. This list is populated by querying kube for master nodes, meaning it's not possible to be in a situation where we don't have a backend.

We are sure to have all 3 master nodes registered with the cluster for etcd operator to be healthy (and appear in the list of backends) before the VIP moves to a master node.

Reply:
I've just pushed [a new commit](https://github.com/openshift/machine-config-operator/pull/3339/commits/339575f7731834b212109626bddb8d5410d25cd2) that should make sure we never end up in a situation like you describe. HAProxy has a way to cause the healthcheck to fail where there is no backends. See https://www.haproxy.com/documentation/hapee/latest/observability/monitoring/monitor-uri/.